blockchain/proxy.go

```diff
@@ -717,6 +743,203 @@ func (ep *proxy) IsDataTrieMigrated(ctx context.Context, address sdkCore.Address
+}
+
+// getBlockNumberByHash retrieves the block number associated with the given block hash
+func (ep *proxy) getBlockNumberByHash(ctx context.Context, shardID uint32, blockHash *[32]byte) (uint64, error) {
+	blockHashStr := hex.EncodeToString(blockHash[:])
+
+	buff, code, err := ep.GetHTTP(ctx, ep.endpointProvider.GetBlockByHash(shardID, blockHashStr))
+	if err != nil || code != http.StatusOK {
+		return 0, createHTTPStatusError(code, err)
+	}
+
+	var response data.BlockResponse
+	if err := json.Unmarshal(buff, &response); err != nil {
+		return 0, err
+	}
+
+	blockNonce := response.Data.Block.Nonce
+
+	// Cache the raw response bytes
+	if ep.filterQueryBlockCacher != nil && len(buff) > 0 {
+		cacheKey := make([]byte, 8)
+		binary.BigEndian.PutUint64(cacheKey, blockNonce)
+		ep.filterQueryBlockCacher.Put(cacheKey, buff, len(buff))
+	}
+
+	return blockNonce, nil
+}
+
+// getLogsFromBlock retrieves logs from a specific block and filters them
+func (ep *proxy) getLogsFromBlock(ctx context.Context, shardID uint32, blockNum uint64, filter *sdkCore.FilterQuery) ([]*transaction.Events, error) {
+	cacheKey := make([]byte, 8)
+	binary.BigEndian.PutUint64(cacheKey, blockNum)
+
+	if ep.filterQueryBlockCacher != nil {
+		cachedResponse, found := ep.filterQueryBlockCacher.Get(cacheKey)
+		if found {
+			cachedBuff, ok := cachedResponse.([]byte)
+			if ok && len(cachedBuff) > 0 {
+				var response data.BlockResponse
+				if err := json.Unmarshal(cachedBuff, &response); err == nil {
+					return extractMatchingEvents(response, filter), nil
+				}
+			}
+		}
+	}
+
+	// Fetch the logs from the network if not found in cache
+	buff, code, err := ep.GetHTTP(ctx, ep.endpointProvider.GetBlockByNonce(shardID, blockNum))
+	if err != nil || code != http.StatusOK {
+		return nil, createHTTPStatusError(code, err)
+	}
+
+	// Cache the raw response bytes
+	if ep.filterQueryBlockCacher != nil && len(buff) > 0 {
+		ep.filterQueryBlockCacher.Put(cacheKey, buff, len(buff))
+	}
+
+	var response data.BlockResponse
+	if err := json.Unmarshal(buff, &response); err != nil {
+		return nil, err
+	}
+
+	if response.Error != "" {
+		return nil, errors.New(response.Error)
+	}
+
+	return extractMatchingEvents(response, filter), nil
+}
+
+func extractMatchingEvents(response data.BlockResponse, filter *sdkCore.FilterQuery) []*transaction.Events {
+	var matchingEvents []*transaction.Events
+	for _, miniblock := range response.Data.Block.MiniBlocks {
+		for _, tx := range miniblock.Transactions {
+			if tx.Logs == nil {
+				continue
+			}
+			for _, event := range tx.Logs.Events {
+				if matchesFilter(filter, event) {
+					matchingEvents = append(matchingEvents, event)
+				}
+			}
+		}
+	}
+	return matchingEvents
+}
+
+func matchesFilter(filter *sdkCore.FilterQuery, event *transaction.Events) bool {
+	// Check if the event's address matches any of the filter addresses (if set)
+	if len(filter.Addresses) > 0 && !contains(filter.Addresses, event.Address) {
```

Comment:
len(filter.Addresses) > 0 can be moved into contains function. Also this function (matchesFilter) would be better to be implemented on the FilterQuery and topicsMatch & contains would be some private methods

Reply:
I would not move this to the FilterQuery struct as that struct will become a hybrid from what it should always be: a DTO

Reply:
len(filter.Addresses) > 0 can be moved into contains function can be done anyway